table/tables/tables.go

```diff
@@ -353,6 +395,27 @@ func (t *Table) getRollbackableMemStore(ctx sessionctx.Context) kv.RetrieverMuta
 	return bs
 }
 
+// locatePartition returns the partition ID of the input record.
+func (t *Table) locatePartition(ctx sessionctx.Context, r []types.Datum) (int64, error) {
+	if len(t.partitionExpr) == 0 {
+		return t.ID, nil
+	}
+
+	for i, expr := range t.partitionExpr {
+		val, _, err := expr.EvalInt(ctx, types.DatumRow(r))
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		if val > 0 {
+			// Partition expression eval to true, the value locates into this partition.
+			partitionInfo := t.meta.GetPartitionInfo()
+			return partitionInfo.Definitions[i].ID, nil
+		}
+	}
+	// The data does not belong to any of the partition?
```

Comment:
We should return an error if data is not in any of the partitions.
Please validate the behavior of MySQL.